extensions/spring-xd-extension-gpfdist/src/main/java/org/springframework/xd/greenplum/gpfdist/GPFDistMessageHandler.java

```diff
@@ -127,49 +134,93 @@ protected void doStart() {
 			context.addLocation(NetworkUtils.getGPFDistUri(gpfdistServer.getLocalPort()));
 
 			sqlTaskScheduler.schedule((new FutureTask<Void>(new Runnable() {
+
 				@Override
 				public void run() {
 					boolean taskValue = true;
 					try {
-						while(!taskFuture.interrupted) {
+						while (!taskFuture.interrupted) {
 							try {
 								greenplumLoad.load(context);
-							} catch (Exception e) {
+							}
+							catch (Exception e) {
 								log.error("Error in load", e);
 							}
-							Thread.sleep(batchPeriod*1000);
+							Thread.sleep(batchPeriod * 1000);
 						}
-					} catch (Exception e) {
+					}
+					catch (Exception e) {
 						taskValue = false;
 					}
 					taskFuture.set(taskValue);
 				}
 			}, null)), new Date());
 
-		} else {
+		}
+		else {
 			log.info("Skipping gpload tasks because greenplumLoad is not set");
 		}
 	}
 
 	@Override
 	protected void doStop() {
+		boolean drained = false;
 		if (greenplumLoad != null) {
+
+			// xd waits 30s to shutdown module, so lets wait 25 to drain
+			long waitDrain = System.currentTimeMillis() + 25000l;
+
+			log.info("Trying to wait buffer to get drained");
+			while (System.currentTimeMillis() < waitDrain) {
+				long capacity = processor.getCapacity();
+				long availableCapacity = processor.getAvailableCapacity();
+				log.info("Buffer capacity " + capacity);
+				log.info("Buffer available capacity " + availableCapacity);
+				if (capacity != availableCapacity) {
+					try {
+						Thread.sleep(1000);
+					}
+					catch (InterruptedException e) {
+					}
+				}
+				else {
+					log.info("Marking stream drained");
+					drained = true;
+					break;
+				}
+			}
+
+			// try to wait current load operation to finish.
 			taskFuture.interruptTask();
 			try {
 				long now = System.currentTimeMillis();
 				// wait a bit more than batch period
+				log.info("Cancelling loading task");
 				Boolean value = taskFuture.get(batchTimeout + batchPeriod + 2, TimeUnit.SECONDS);
 				log.info("Stopping, got future value " + value + " from task which took "
 						+ (System.currentTimeMillis() - now) + "ms");
-			} catch (Exception e) {
+			}
+			catch (Exception e) {
 				log.warn("Got error from task wait value which may indicate trouble", e);
```

Comment:
I just wanted to add a logging here instead of just ignoring if we get an exception from a `TaskScheduler` when cancelling `Future`. I'm not entirely sure when that exception would be thrown but wanted to add it to logs in case shutdown somehow still fails and we're getting exception here.